File: src/shrinker/integer.ts
import { Random } from '../Random'
import { Stream } from '../Stream'
import { Shrinkable } from '../Shrinkable'

// Generates a stream of shrink candidates for a positive integer range using binary search.
// It repeatedly splits the range [min, max) in half, shrinking towards min.
function genpos(min: number, max: number): Stream<Shrinkable<number>> {
    // Calculate the midpoint, ensuring it rounds towards min correctly, even for odd ranges.
    const mid = Math.floor(min / 2) + Math.floor(max / 2) + (min % 2 !== 0 && max % 2 !== 0 ? 1 : 0)
    if (min + 1 >= max) return Stream.empty() // Base case: No more shrinking possible.
    else if (min + 2 >= max) return Stream.one(new Shrinkable<number>(mid)) // Base case: Only the midpoint is left.
    // Recursively generate shrinks: prioritize the midpoint, then the lower half, then the upper half.
    else return Stream.one(new Shrinkable<number>(mid).with(() => genpos(min, mid))).concat(genpos(mid, max))
}

// Generates a stream of shrink candidates for a negative integer range using binary search.
// It repeatedly splits the range (min, max] in half, shrinking towards max.
function genneg(min: number, max: number): Stream<Shrinkable<number>> {
    // Calculate the midpoint, ensuring it rounds towards max correctly, even for odd ranges.
    const mid = Math.floor(min / 2) + Math.floor(max / 2) + (min % 2 !== 0 && max % 2 !== 0 ? -1 : 0)
    if (min + 1 >= max) return Stream.empty() // Base case: No more shrinking possible.
    else if (min + 2 >= max) return Stream.one(new Shrinkable<number>(mid)) // Base case: Only the midpoint is left.
    // Recursively generate shrinks: prioritize the midpoint, then the lower half (closer to max), then the upper half.
    // Note: genpos is used for the lower half because it shrinks towards the lower bound (which is closer to 0).
    else return Stream.one(new Shrinkable<number>(mid).with(() => genpos(min, mid))).concat(genpos(mid, max))
}

/**
 * Creates a Shrinkable<number> that shrinks towards 0 using a binary search approach.
 *
 * @param value The initial integer value.
 * @returns A Shrinkable number that shrinks towards 0.
 */
export function binarySearchShrinkable(value: number): Shrinkable<number> {
    return new Shrinkable<number>(value).with(() => {
        if (value === 0) return Stream.empty() // 0 cannot shrink further.
        // For positive numbers, shrink towards 0: prioritize 0, then use genpos for the range (0, value).
        else if (value > 0) return Stream.one(new Shrinkable<number>(0)).concat(genpos(0, value))
        // For negative numbers, shrink towards 0: prioritize 0, then use genneg for the range (value, 0).
        else return Stream.one(new Shrinkable<number>(0)).concat(genneg(value, 0))
    })
}

/**
 * Generates a Shrinkable integer within the specified range [min, max].
 * The generated value shrinks towards the minimum bound if min >= 0,
 * towards the maximum bound if max <= 0, or towards 0 otherwise.
 *
 * @param random The random number generator.
 * @param min The minimum inclusive bound.
 * @param max The maximum inclusive bound.
 * @returns A Shrinkable integer within the specified range.
 * @throws {Error} If the generated value is outside the [min, max] range (should not happen with correct random.intInterval).
 */
export function generateInteger(random: Random, min: number, max: number): Shrinkable<number> {
    const value = random.intInterval(min, max)
    // Basic range check for safety, though random.intInterval should guarantee this.
    if (value < min || max < value) throw new Error('invalid range')

    // If the range is entirely non-negative, shrink towards min.
    if (min >= 0) return binarySearchShrinkable(value - min).map(n => n + min)
    // If the range is entirely non-positive, shrink towards max.
    else if (max <= 0) return binarySearchShrinkable(value - max).map(n => n + max)
    // If the range crosses zero, shrink towards 0.
    else return binarySearchShrinkable(value)
}

// TODO: shrinkIntegral

File: dartproptest/lib/src/shrinker/integer.dart
import '../random.dart';
import '../shrinkable.dart';
import '../stream.dart';

/// Generates a shrinkable integer within the specified range with shrinking capabilities.
///
/// [random] The random number generator.
/// [min] The minimum value (inclusive).
/// [max] The maximum value (inclusive).
/// Returns a Shrinkable containing the generated integer and its shrinks.
Shrinkable<int> generateInteger(Random random, int min, int max) {
  final value = random.interval(min, max);
  return Shrinkable<int>(value, () => _shrinkInteger(value, min, max));
}

/// Generates shrinks for an integer value.
///
/// [value] The integer value to shrink.
/// [min] The minimum allowed value.
/// [max] The maximum allowed value.
/// Returns a stream of smaller integer values.
LazyStream<Shrinkable<int>> _shrinkInteger(int value, int min, int max) {
  if (value == min) {
    return LazyStream<Shrinkable<int>>(null);
  }

  // Generate shrinks by halving the distance to min
  final List<Shrinkable<int>> shrinks = [];

  // Try shrinking towards zero first
  if (value > 0 && min <= 0) {
    shrinks.add(Shrinkable<int>(0));
  }

  // Try shrinking towards min
  if (value > min) {
    final newValue = min + (value - min) ~/ 2;
    if (newValue != value && newValue >= min) {
      shrinks.add(
          Shrinkable<int>(newValue, () => _shrinkInteger(newValue, min, max)));
    }
  }

  // Try shrinking by removing the last digit (for positive numbers)
  if (value > 0) {
    final newValue = value ~/ 10;
    if (newValue != value && newValue >= min) {
      shrinks.add(
          Shrinkable<int>(newValue, () => _shrinkInteger(newValue, min, max)));
    }
  }

  // Try shrinking by removing the last digit (for negative numbers)
  if (value < 0) {
    final newValue = value ~/ 10;
    if (newValue != value && newValue <= max) {
      shrinks.add(
          Shrinkable<int>(newValue, () => _shrinkInteger(newValue, min, max)));
    }
  }

  // Convert list to stream
  if (shrinks.isEmpty) {
    return LazyStream<Shrinkable<int>>(null);
  }

  return LazyStream<Shrinkable<int>>(shrinks.first, () {
    if (shrinks.length == 1) {
      return LazyStream<Shrinkable<int>>(null);
    }
    return LazyStream<Shrinkable<int>>(shrinks[1], () {
      if (shrinks.length == 2) {
        return LazyStream<Shrinkable<int>>(null);
      }
      return LazyStream<Shrinkable<int>>(shrinks[2]);
    });
  });
}

/// Creates a shrinkable that uses binary search to shrink a range.
/// This is useful for shrinking array lengths and other range-based values.
///
/// [range] The range to shrink (e.g., array length - minSize).
/// Returns a Shrinkable that shrinks the range using binary search.
Shrinkable<int> binarySearchShrinkable(int range) {
  if (range <= 0) {
    return Shrinkable<int>(0);
  }

  return Shrinkable<int>(range, () => _binarySearchShrinks(range));
}

/// Generates binary search shrinks for a range.
LazyStream<Shrinkable<int>> _binarySearchShrinks(int range) {
  if (range <= 0) {
    return LazyStream<Shrinkable<int>>(null);
  }

  final half = range ~/ 2;
  if (half == 0) {
    return LazyStream<Shrinkable<int>>(null);
  }

  return LazyStream<Shrinkable<int>>(
      Shrinkable<int>(half, () => _binarySearchShrinks(half)));
}
